src/discord.cpp

```diff
@@ -0,0 +1,237 @@
+    }
+}
+
+void Discord::handleDiscordReady(const DiscordUser* request)
+{
+    Q_UNUSED(request);
+
+    mudlet::self()->mDiscord.UpdatePresence();
+}
+
+void Discord::handleDiscordDisconnected(int errorCode, const char* message)
+{
+    qWarning() << "Discord disconnected:" << errorCode << message;
+}
+
+void Discord::handleDiscordError(int errorCode, const char* message)
+{
+    qWarning() << "Discord error:" << errorCode << message;
+}
+
+void Discord::handleDiscordJoinGame(const char* joinSecret)
+{
+    Q_UNUSED(joinSecret);
+}
+
+void Discord::handleDiscordSpectateGame(const char* spectateSecret)
+{
+    Q_UNUSED(spectateSecret);
+}
+
+void Discord::handleDiscordJoinRequest(const DiscordUser* request)
+{
+    Q_UNUSED(request);
+}
+
+void Discord::UpdatePresence()
+{
+    DiscordRichPresence discordPresence;
+    memset(&discordPresence, 0, sizeof(discordPresence));
+
+    char buffer[256];
+    buffer[0] = '\0';
+
+    auto host = mudlet::self()->getActiveHost();
+    if (!host) {
+        return;
+    }
+
+    auto gameName = mGamesNames[host].toUtf8();
+    auto gameNameLowercase = mGamesNames[host].toLower().toUtf8();
+    auto area = mAreas[host].toUtf8();
+    auto characterIcon = mCharacterIcons[host].toLower().toUtf8();
+    auto characterText = mCharacters[host].toUtf8();
+    auto url = host->getUrl();
+    auto port = QString::number(host->getPort());
+
+    // list of known games - if the user sets an unknown game, have a heuristic
+    // still display their character icon
+    bool knownGame = mKnownGames.contains(gameNameLowercase);
+
+    if (!gameName.isEmpty()) {
+        sprintf(buffer, "Playing %s", gameName.constData());
+        discordPresence.details = buffer;
+        discordPresence.largeImageKey = gameNameLowercase.constData();
+
+        if (!host->mDiscordHideAddress) {
+            discordPresence.largeImageText = QStringLiteral("%1:%2").arg(url, port).toUtf8().constData();
+        }
+    }
+
+    if (!area.isEmpty()) {
+        discordPresence.state = area.constData();
+    }
+
+    if (!characterIcon.isEmpty()) {
+        // the game is unknown, set the small image as the big one so at least something shows
+        if (knownGame) {
+            discordPresence.smallImageKey = characterIcon.constData();
+        } else {
+            discordPresence.largeImageKey = characterIcon.constData();
+        }
+    }
+
+    if (!characterText.isEmpty()) {
+        if (knownGame) {
+            discordPresence.smallImageText = characterText.constData();
+        } else {
+            if (!host->mDiscordHideAddress) {
+                discordPresence.largeImageText = QStringLiteral("%1:%2 | %3").arg(url, port, characterText).toUtf8().constData();
+            } else {
+                discordPresence.largeImageText = characterText.constData();
+            }
+        }
+    }
+
+//    discordPresence.startTimestamp = mStartTime;
+
+    discordPresence.instance = 1;
+    Discord_UpdatePresence(&discordPresence);
```

Comment:
☠️ I get a Seg. Fault here (Mudlet crashes and burns) because I checked on one of the "Special Options" -> "Discord privacy" checkboxes in the "Profile preferences" - obviously this is because my `DiscordRichPresence` instance `discordPresence` is a *big box of null pointers* on my OS Platform (although the `(int8_t) instance` does have the value of `1` at the point of crashing!)

We need to provide a means to disable this functionality either for user choice (I was iffy about T&Cs initially remember!) or for packagers (I'm not sure that Discord meets the [**DFSG**](https://en.wikipedia.org/wiki/Debian_Free_Software_Guidelines) so it may make it awkward for @csmall to include it in the Debian package) or for porters/platforms where the Discord application **cannot** be run. {Besides FreeBSD, Vista is not supported though it only reach EOL last year and I'm not sure there is a 32-Bit Linux version of the Discord App even if there are third party builds of Mudlet for that...}

Reply:
Fixed, let me know if it's an issue now.